Add new guide on how to update to aws provider v3 (#370)
gruntwork-io/gruntwork-io.github.io: _posts/2020-10-22-how-to-update-to-aws-provider-v3.adoc

---
title: How to update Gruntwork references to Version 3 of the Terraform AWS provider
categories: Foundations

> image: /assets/img/guides/infrastructure-as-code-library/grunty-grid.png

Author: I'm not sure what is the best icon to fit here. Used grunty for now because this is very specific to the Gruntwork products, but maybe there is something better to use?

Reviewer: Perhaps a 3.0 icon? Or refresh icon? Or update badge icon?

---

> title: How to update Gruntwork references to Version 3 of the Terraform AWS provider

Reviewer: ```suggestion
title: How to update to version 3.x of the Terraform AWS Provider
```

> categories: Foundations

Reviewer: Perhaps we should have an upgrade-guides category?

image: /assets/img/guides/infrastructure-as-code-library/grunty-grid.png

> excerpt: Learn how to update the Gruntwork Reference Architecture and your infrastructure code that depends on the Gruntwork Infrastructure as Code Library to use the latest version of the Terraform AWS provider.

Reviewer: ```suggestion
excerpt: Learn how to update the Gruntwork Reference Architecture and your infrastructure code that depends on the Gruntwork Infrastructure as Code Library to use version 3.x of the Terraform AWS provider.
```

tags: ["aws", "terraform", "terragrunt"]
cloud: ["aws"]
redirect_from: /static/guides/foundations/how-to-update-to-aws-provider-v3/
---
:page-type: guide
:page-layout: post

:toc:
:toc-placement!:

// GitHub specific settings. See https://gist.github.com/dcode/0cfbf2699a1fe9b46ff04c41721dda74 for details.
ifdef::env-github[]
:tip-caption: :bulb:
:note-caption: :information_source:
:important-caption: :heavy_exclamation_mark:
:caution-caption: :fire:
:warning-caption: :warning:
toc::[]
endif::[]

== Intro

This guide will walk you through how to update the https://gruntwork.io/reference-architecture/[Gruntwork Reference
Architecture] and any code that depends on the

> https://gruntwork.io/infrastructure-as-code-library/[Gruntwork Infrastructure as Code Library] to the latest version
> (v3) of the Terraform AWS provider.

Reviewer: ```suggestion
https://gruntwork.io/infrastructure-as-code-library/[Gruntwork Infrastructure as Code Library] to the version 3.x of the Terraform AWS provider.
```

It won't be "latest" forever.

Reviewer: Also, consider adding a sentence here as to _why_ this matters: e.g., version 3.x has lots of new features (link), but also a number of backwards incompatible changes that have to be incorporated carefully into your codebase.

Author: Well, version 3.x doesn't really have a single page I can link to with "new features" (there are 12 minor releases, each with "new" stuff)... The real thing here is that 2.x is deprecated and nothing new will be going in there.

This is why I went for the following phrasing in the section below:

> This means that if you wish to use any new resources or data sources that are released from this point onwards, you will need to be compatible with version 3 of the AWS provider to be able to pull those changes in.

I'll copy a similar statement to the intro.

Author: Done in 6ea3831db7e56d9bb73ff6e6d0697e500690f24d

=== What you'll learn in this guide

This guide consists of three main sections:

<<core_concepts>>::
  An overview of version 3 of the AWS provider and why it is important to update your code for compatibility.

<<deployment_walkthrough>>::
  The steps you need to take to update your code relying on the Gruntwork Infrastructure as Code library and your

>   version of the Gruntwork Reference Architecture with compatibility with AWS provider v3.

Reviewer: NIT: mention the compatibility table

Author: Done in f1dfdb0cd96a753314e5bba1a0a4ced40a78ed81

[[core_concepts]]
== Core Concepts

=== Background

Version https://github.com/terraform-providers/terraform-provider-aws/releases/tag/v3.0.0[v3.0.0] of the Terraform AWS

> provider was released on July 30th 2020 with backwards incompatible updates. Following the provider release cycle,

Reviewer: NIT: mention the new items it brings too! It's not all bad changes 😁

Author: There isn't a good summary of the new items it brings, which is why I went for the specific phrasing in the statement right afterwards.

future releases from this point onward would only be compatible with 3.X. This means that if you wish to use any new
resources or data sources that are released from this point onwards, you will need to be compatible with version 3 of
the AWS provider to be able to pull those changes in.

Since this is a major version change, there are a number of backwards incompatible changes that were introduced. In
order to be able to use version 3, you need to update the resources and data sources with backwards incompatible

> changes.

Reviewer: NIT: it's not so much the resources / data sources you need to update, but _your code_ and _your Terraform state file_.

Author: NOTE: There are no terraform state changes in this update if it is just the aws provider changes (there is a bunch in the ref arch because we are jumping 5-10 backwards incompatible changes from our modules).

I'll rephrase this, but I was really going for "update your resource and data source blocks".

Author: Rephrased in e198cac59996c0af5e0f57978b119abbb95b7c4e

For your own code, you can follow
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-3-upgrade[the official guide] on
updating to AWS provider version 3. The rest of this guide will focus on updating references to Gruntwork modules and
the Gruntwork Reference Architecture to be compatible with AWS provider version 3.

[[deployment_walkthrough]]
== Deployment walkthrough

=== Updating references to Gruntwork Infrastructure as Code Library

In order to take advantage of the Terraform AWS provider version 3, you need to update your references to the Gruntwork
Infrastructure as Code Library to use a compatible version. The following table provides a summary of all the relevant

> Gruntwork AWS modules and the respective versions that are compatible with AWS provider version 3.

Reviewer: You might want to mention here, and perhaps in the intro, that we (Gruntwork) did the hard work of testing and updating our whole library and Ref Arch to be compatible with AWS Provider 3.x. As a customer, you need to update to the proper versions of the Gruntwork library to pick up the fixes/changes we had to make, plus make some code and state changes in your own codebase.

Reviewer: Also, after this paragraph, add an example of how to update. E.g., Look for code with `module { source = "github.com/gruntwork-io/xxx//yyy?ref=zzz" }`, bump `ref=zzz` to the corresponding version number below. And make sure to check out the release notes for those releases for migration guides!

Author: Reworded to include that we tested and updated the library: 259d8b08d0240ad8511b95945b3afc89ceb9440d

I didn't include the phrase "hard work", because IMHO, it feels a bit too pretentious and arrogant.

Also updated to link to "Updating" section of the "how to use Gruntwork" guide: ced5bfbe19a95becdb71daf97e128da9c1f9e0ae

[.exceptional]
IMPORTANT: Gruntwork follows semantic versioning. In practice this means that version updates to the minor version are
considered backwards incompatible releases for any version updates prior to 1.0.0 release. Make sure to read the release

> notes for the relevant modules any time you are updating minor versions!

Reviewer: I think this is missing some important details. You may want to cross-link to https://gruntwork.io/guides/foundations/how-to-use-gruntwork-infrastructure-as-code-library/#versioning instead.

Author: Hmm... I just took a look through the `versioning` section again and I feel like I captured the most relevant detail here. I know I missed some of the overviews, but I think just linking to versioning is not sufficient, because:

- I suspect most users won't click the link or read it carefully (probably will only skim). You are more likely to read the thing in the red box in its entirety, than a link in the red box, especially if it is not written in the same context.
- The most important detail about reading the release notes is buried in the sea of text.

I added a link and reworded it to highlight the most relevant detail from that section to bring it to context to address the 2 concerns I had above.

259d8b08

[cols="1,1h"]
|===

> |IaC Package |Minimum version with AWS Provider v3 support

Reviewer: ```suggestion
|Gruntwork Repo |Minimum version with AWS Provider v3 support
```

Reviewer: Did we track whether each release required migration steps? Would be great to have a column like, "requires migration steps" and for the ones that require, say, code changes or state changes, we say, "Yes, see migration guide (link)."

Author: None of them require actual migrations to the code (the increments are all backwards compatible). However, that is only true if you are going to that version from the immediate previous version. This was why I mentioned the "IMPORTANT" note above.

> |module-vpc

Reviewer: ```suggestion
|terraform-aws-vpc
```

> |v0.9.0

Reviewer: Make these links to the release notes.

Author: Done in 986b757a0249a2f60a55d04bfdcfe5325a1b10b9

|module-asg
|v0.10.0

|module-data-storage
|v0.15.0

|module-server
|v0.8.5

|package-lambda
|v0.8.1

|package-sam
|v0.2.1

|module-cache
|v0.9.4

|package-messaging
|v0.3.4

|package-static-assets
|v0.6.5

> |module-aws-monitoring

Reviewer: ```suggestion
|terraform-aws-monitoring
```

|v0.22.1

|package-openvpn
|v0.11.0

|module-security
|v0.35.0

|cis-compliance-aws
|v0.6.0

|module-ecs
|v0.22.0

|module-ci
|v0.27.3

|terraform-aws-eks
|v0.22.1

|module-load-balancer
|v0.20.4

|terraform-aws-couchbase
|v0.3.0

|package-zookeeper
|v0.6.7

|package-kafka
|v0.6.3

|package-elk
|v0.6.0

|terraform-aws-influx
|v0.1.3

|terraform-aws-vault
|v0.13.11

|terraform-aws-consul
|v0.7.10

|terraform-aws-nomad
|v0.6.6

|package-terraform-utilities
|v0.2.1

|===

=== Updating the Gruntwork Reference Architecture to AWS Provider v3

If you purchased the Gruntwork Reference Architecture, you will have a copy of the `infrastructure-live` and

> `infrastructure-modules` repositories that contain the infrastructure code for deploying the Reference Architecture. You
> will need to follow the steps to update the relevant code in `infrastructure-modules` to use a compatible version of the
> Gruntwork Infrastructure as Code Library, following the steps described above.

Reviewer: ```suggestion
`infrastructure-modules` repositories that contain the infrastructure code for deploying the Reference Architecture. You
will need to update the relevant code in `infrastructure-modules` to use a compatible version of the
Gruntwork Infrastructure as Code Library, as per the compatibility table above.
```

(consider making compatibility table a cross-link)

Author: Done in 97eb4617437e9bdd7270b9471260c4dde149764f

> To help guide through the upgrade process, we have updated the Acme Reference Architecture examples to support AWS

Reviewer: ```suggestion
To help guide you through the upgrade process, we have updated the Acme Reference Architecture examples to support AWS
```

provider v3. You can refer to the following release notes for detailed information and code patches that you can apply
to update your snapshot of the Gruntwork Reference Architecture:

> CIS flavor Reference Architecture::

Reviewer: ```suggestion
CIS Reference Architecture::
```

  Refer to the release notes for
  https://github.com/gruntwork-io/cis-infrastructure-live-acme/releases/tag/v0.0.1-20201021[v0.0.1-20201021] of the
  `cis-infrastructure-live-acme` repository for instructions on how to update the CIS components of the Reference
  Architecture to be compatible with AWS provider v3. For all other components, refer to the release notes for
  https://github.com/gruntwork-io/infrastructure-modules-multi-account-acme/releases/tag/v0.0.1-20201021[v0.0.1-20201021]
  of the `infrastructure-modules-multi-account-acme` repository.

> Standard flavor Reference Architecture::

Reviewer: ```suggestion
Standard Reference Architecture::
```
